mailpoet/lib/Form/DisplayFormInWPContent.php

```diff
@@ -187,8 +187,11 @@ private function getContentBellow(FormEntity $form, string $displayType): string
       $templateData['enableExitIntent'] = true;
     }
 
+    $uniqueFormAction = Security::generateRandomString();
+    $templateData['mailpoetAction'] = $uniqueFormAction;
```

Comment:
It [seems that WP suggests](https://codex.wordpress.org/WordPress_Nonces) simply using IDs such as form ID in our case and I almost wanted to ask why we don't do that as well... but then I realized that we need the randomness for the guest users.

But that made me think... is it really correct to do this for guests? Many sites use CDNs, many pages may be cached, and when there's no session, the cache will break this mechanism. I'm not sure if this can work for guests at all and I just found WC was discussing the same: https://github.com/woocommerce/woocommerce/issues/2921

This makes me think whether the issue we're trying to solve is a valid issue. It doesn't seem so — nonces can prevent an attacker from doing a request on your behalf (e.g. when you're an admin) without you knowing that. But when you're a guest we don't care what they do on "your behalf" as you have no specific session tied to you.

I'm thinking this issue may be invalid in the first place. What do you think?

Reply:
I agree that the issue as stated is probably invalid to begin with. I also know that sometimes these issues will continue to be flagged by security scanners and sometimes it's easier to change the code than to have to explain why a result is invalid. If we don't have any kind of obligation to prevent these kinds of issues (i.e. ones that aren't truly a problem) from coming up in scans I'm happy to avoid it.

Perhaps the reason why we were ignoring the token if the captcha is turned on is to avoid the caching issue?

I did also consider using form ID, or some kind of hash based on form data, and part of the reason I moved away from that was that I didn't want to clutter up `setupAjax` with more logic that's specific to a particular endpoint. That was a big part of the reason I wanted to get rid of the `$ignoreToken` logic

```php
    $ignoreToken = (
      $this->settings->get('captcha.type') != Captcha::TYPE_DISABLED &&
      $this->requestEndpoint === 'subscribers' &&
      $this->requestMethod === 'subscribe'
    );
```

We were including the `SettingsController` and had `use MailPoet\Subscription\Captcha;` specifically for one endpoint. I didn't want to have to import other classes just to retrieve the form if we wanted to generate an action based on more than the form ID. It might seem silly or like overkill to retrieve the form entity for the nonce, but one of the potential issues I was thinking about is this:

1) A form is created for segments 1 and 2
2) The form is rendered on a page and gets cached
3) The form is changed to subscribe users to segments 3, 4, and 5
4) A user sees the cached form and submits it

Would we still want the form to work if the segments changed? Users could potentially be signing up for something they weren't intending to. If we calculate an action based on the segment IDs then we could avoid that problem, though I don't know if it truly is a problem.

Overall I was wondering if it would make sense to reconsider how we use nonces in general. Perhaps it makes sense to hand over the responsibility of generating/checking nonces to the request endpoint classes themselves, since they would be better equipped to create action-specific actions instead of us always using `mailpoet_token`. The effort wouldn't be trivial though and I don't know that we would gain much from it.

Truly the biggest potential issue I see with our current implementation is that a person could more easily re-use a specific form by making direct API requests with `curl` or similar, and that's only an issue if captchas are disabled, which they aren't by default.

Anyway, that's a lot of text mostly to say that I agree that the original issue isn't valid as such, and maybe it would be best to change nothing at all because we're better off ignoring the tokens when captchas are enabled to avoid issues with cached forms.

Reply:
> sometimes it's easier to change the code than to have to explain why a result is invalid

I think the problem with this ticket, in particular, is that if we implement it, we will break forms on cached pages and it will happen in two ways (no matter whether captcha is used or not):

1. If a page is still in the cache and the nonce is expired, the form will not work.
2. If a user caches a nonce and we have the one-use policy, it will not work for another user (from the cache).

Additionally, the cache would still re-use that same nonce for multiple users (and we would have to make remove the one-use logic).

Therefore, it seems to me that we can't implement this as we would essentially create new problems with forms.

> Would we still want the form to work if the segments changed? Users could potentially be signing up for something they weren't intending to.

I'd say yes as this is a caching issue, not a security issue. I'm not sure what the form includes (if list IDs or just form ID) but it's something we can't really avoid anyway (a nonce would be cached too).

> Overall I was wondering if it would make sense to reconsider how we use nonces in general.

I agree, seems like we should use action-specific nonces, I'm not sure how often we use them now. But at the same time, it seems to me that for a REST API it's used like this: https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/

> Anyway, that's a lot of text mostly to say that I agree that the original issue isn't valid as such, and maybe it would be best to change nothing at all because we're better off ignoring the tokens when captchas are enabled to avoid issues with cached forms.

Yes, I'm afraid we can't merge it anyway, as mentioned above. I don't see captcha to be very related btw., as the nonces are for CSRF protection while captcha is against spambots. CSRF is dangerous when you're authenticated as someone could make your browser do unwanted actions, but doesn't seem to make any sense for guest users.

Reply:
> If a page is still in the cache and the nonce is expired, the form will not work.

Isn't this also true currently? I suppose only when captchas are disabled, and I suspect that's a very uncommon configuration.

> I'm not sure what the form includes (if list IDs or just form ID) but it's something we can't really avoid anyway (a nonce would be cached too).

In this case I was thinking we wouldn't include the segment IDs in the form itself, rather we would use those IDs to generate the nonce and also to check it, which would all be server side. For example, the nonce action might be `mailpoetSubscribeForm1Segments1,2`. At the time of the check it would be `mailpoetSubscribeForm1Segments3,4,5`. But like you said it's a totally separate issue.

> I agree, seems like we should use action-specific nonces, I'm not sure how often we use them now. But at the same time, it seems to me that for a REST API it's used like this: https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/

Thanks for pointing this out. If I'm understanding it correctly WP doesn't even check (or care) about nonces for logged out users, is that right? It makes sense since, as you've pointed out, CSRF protection doesn't really make much sense for logged out users.

Is there anything you think would be worthwhile to do at this point? Perhaps just getting rid of `Security`'s `generateToken` method and using our `WPFunctions` wrapper directly instead? We could also ignore tokens for all unauthenticated requests to match WordPress's behavior and avoid caching issues.

Reply:
> Isn't this also true currently? I suppose only when captchas are disabled, and I suspect that's a very uncommon configuration.

I don't know. Is caching related to whether captchas are used? Is the captcha included on the first load or as an intermediate step after submit? I don't really remember how it works.

Seeing that WC went the same way to not using the tokens for guest sessions, I guess we could do the same.

> Is there anything you think would be worthwhile to do at this point?

I guess we can do the improvements you're suggesting. Maybe it would make the code more explicit about this if we ignore the tokens.

Reply:
> I don't know. Is caching related to whether captchas are used? Is the captcha included on the first load or as an intermediate step after submit? I don't really remember how it works.

What I meant is that it would only be an issue with captchas disabled because we're ignoring tokens in that case. If there's a cached form and the token is > 24 hours old, the nonce check will fail.